source/agora/consensus/BlockRewardCalc.d

```diff
@@ -0,0 +1,650 @@
+
+    Params:
+        height = the height for which we want to get the period information
+        block_rewards = time periods
+
+    Returns: which 'period' the particular height falls into and whether that height
+    is the last height for that particular period
+
+    ***************************************************************************/
+
+    private CurrentPeriodTup getCurrentPeriod (in Height height, immutable(BlockRewardsTup)[] block_rewards) @safe const pure nothrow
+    {
+        ulong blocks_accumulated = 0;
+        foreach (period; 0 .. block_rewards.length)
+        {
+            blocks_accumulated += block_rewards[period].dist_period_secs / consensus_params.BlockInterval.total!"seconds";
+            if (blocks_accumulated >= height)
+                return CurrentPeriodTup(period, blocks_accumulated == height);
+        }
+        return CurrentPeriodTup(block_rewards.length, false);
+    }
+
+    unittest
+    {
+        immutable block_rewards = [BlockRewardsTup(Amount(100), 7), BlockRewardsTup(Amount(100), 2)];
+        auto block_reward_calc = getBlockRewardCalc(2, 5, block_rewards);
+        assert(block_reward_calc.getCurrentPeriod(Height(0), block_rewards) == CurrentPeriodTup(0, false));
+        assert(block_reward_calc.getCurrentPeriod(Height(7), block_rewards) == CurrentPeriodTup(0, true));
+        assert(block_reward_calc.getCurrentPeriod(Height(8), block_rewards) == CurrentPeriodTup(1, false));
+        assert(block_reward_calc.getCurrentPeriod(Height(9), block_rewards) == CurrentPeriodTup(1, true));
+        assert(block_reward_calc.getCurrentPeriod(Height(10), block_rewards) == CurrentPeriodTup(2, false));
+    }
+
+    /***************************************************************************
+
+    Returns the validator reward for each validator based on how many blocks
+    each validator signed, and also calculates the remainder that cannot be evenly
+    distributed to the validators.
+
+    Params:
+        height = the height of the block that will contain the CoinBase transaction
+        validator_stakes = number of blocks signed by the validator identified by
+                           its public key
+        remainder = the remainder that cannot be evenly distributed to
+                    the validators
+        comply_perc = (#actual signatures / #expected signatures) * 100 on all
+              the blocks for which we want to calculate the block reward
+        indent = number of spaces before log messages
+
+    Returns: the validator reward for each validator based on how many blocks
+    each validator signed
+
+    ***************************************************************************/
+
+    public Amount[Point] getEachValidatorReward (in Height height, ushort[Point] validator_stakes,
+        out Amount remainder, double comply_perc, uint indent = 0) @safe const nothrow
+    {
+        auto indentStr = ' '.repeat(indent).array;
```

Comment:
This will cause an allocation even if trace logging is disabled. Could perhaps turn it into a function?

Reply:
> This will cause an allocation even if trace logging is disabled. Could perhaps turn it into a function?

hmm, Ocean's format function(called by log.trace) is defined as 
```D
public void format (Args...) (Level level, cstring fmt, Args args) @safe
```
so I cannot just lazily evaluate the argument.

the only thing I can think of is 
```D
class IndentClass(int indentation)
{
    void toString (scope FormatterSink sink)
    {
        sink(' '.repeat(indent).array);
    }
}
```

and then create an instance of this class on the stack, and then pass that instance to log.trace :) 

do you think it is worth the effort?

Reply:
Maybe it's overkill then. Let's fix it later if it's not a big deal.

Reply:
There's a simple fix: https://github.com/sociomantic-tsunami/ocean/blob/ba566c473249341633d723904a0b82ba1201a55b/src/ocean/util/log/Logger.d#L385-L391